go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+# Go SDK Workflow Interceptor Updates and OpenTelemetry Tracing Support
+
+This proposal is for updates to Go workflow interceptors with the following features:
+
+* Support for and implementation of OpenTelemetry tracing
+* Support idiomatic Go interceptor design
+* Provide feature parity with Java SDK and other missing features (see https://github.com/temporalio/sdk-go/issues/529)
+* Add activity interceptors
+* Make inbound interceptors future-proof for argument addition
+
+**These updates are known to NOT be backwards-compatible**
+
+## Interceptor Proposed Updates
+
+### WorkflowInterceptor.InterceptWorkflow
+
+Current:
+
+```go
+InterceptWorkflow(info *WorkflowInfo, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Proposal:
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Why: Callers can use the context to get workflow info.
+
+### WorkflowXCallsInterceptor
+
+Proposal: Change the name to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more accepted in Java, not Go).
+
+### WorkflowInboundCallsInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
+```
+
+Why:
+
+* Workflow type can be derived from the context _or_ the info first given to `InterceptWorkflow`
```

Comment:
Where can we access the headers? in the context? In the other SDKs it's a part of the input.
Also `workflowType` is part of the input and may be modified by an interceptor (although I'm not sure this is desirable).

Reply:
> Where can we access the headers? in the context?

Headers are usually part of the context in Go, yes. But I can explicitly confirm that behavior.

> Also workflowType is part of the input and may be modified by an interceptor (although I'm not sure this is desirable).

I am unable to find this behavior in the Java SDK. Do we want this feature here and there?

Reply:
It doesn't exist in `WorkflowInput` although `SignalInput` does include `signalName`:
https://github.com/temporalio/sdk-java/blob/6f86b160d9b948418d2e49d8e5a2bbb260fad149/temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowInboundCallsInterceptor.java#L61

I can imagine a use-case where based on some logic user decided to re-route to a different workflow or signal.


Are you referring to [Value / withValue](https://pkg.go.dev/context#WithValue)?
API headers strictly map from `string` to `Payload`, should we enforce this?

Reply:
After further discussion, while I will still be adding `WithHeaders`, the `HeaderProvider` concept cannot be deprecated as that is the primary way to add headers via gRPC interceptor. The client/activity/workflow interceptors do not map 1:1 w/ gRPC requests (we even make requests outside of intercepted calls like for registering namespace).